doc/specs/#2046 - Command Palette.md

```diff
@@ -0,0 +1,284 @@
+name, to show how their input correlates to the results on screen.
+
+## Capabilities
+
+### Accessibility
+
+As the entire command palette will be a native XAML element, it'll automatically
+be hooked up to the UIA tree, allowing for screen readers to naturally find it.
+ * When the palette is opened, it will automatically recieve focus.
+ * The terminal panes will not be able to be interacted with while the palette
+   is open, which will help keep the UIA tree simple while the palette is open.
+
+### Security
+
+This should not introduce any _new_ security concerns. We're relying on the
+security of jsoncpp for parsing json. Adding new keys to the settings file
+will rely on jsoncpp's ability to securely parse those json values.
+
+### Reliability
+
+We'll need to make sure that invalid commands are ignored. A command could be
+invalid because:
+* it has a null `name`, or a name with the empty string for a value.
+* it has a null `action`, or an action specified that's not an actual
+  `ShortcutAction`.
+
+We'll ignore invalid commands from the user's settings, instead of hard
+crashing. I don't believe this is a scenario that warrants an error dialog to
+indicate to the user that there's a problem with the json.
+
+### Compatibility
+
+We will need to define default _commands_ for all the existing keybinding
+commands. With #754, we could add all the actions (that make sense) as commands
+to the commands list, so that everyone wouldn't need to define them manually.
+
+### Performance, Power, and Efficiency
+
+We'll be adding a few extra XAML elements to our tree which will certainly
+increase our runtime memory footprint while the palette is open.
+
+We'll aditionally be introducing a few extra json values to parse, so that could
+increase our load times (though this will likely be negligible).
+
+## Potential Issues
+
+This will first require the work in [#1205] to work properly. Right now we
+heavily lean on the "focused" element to determine which terminal is "active".
+However, when the command plaette is opened, focus will move out of the terminal
+control into the command palette, which leads to some hard to debug crashes.
+
+Additionally, we'll need to ensure that the "fuzzy search" algorithim proposed
+above will work for non-english languages, where a single charater might be
+multiple `char`s long. As we'll be using a standard XAML text box for input, we
+won't need to worry about handling the input ourselves.
+
+## Future considerations
+
+* Once [#1142] is also complete, we'll also add a `executeCommand` action. It
+  will take a single parameter `command`, which will be the name of a command
+  from the list of commands. This will enable a user to bind a command to a
+  keybinding. This could allow the user to bind a specific pairing of `action`
+  and `args` that they've set up in a command as a keybinding, without having to
+  repeat the entry in both the `commands` and `keybindings`.

```

Comment:
scary. i don't like the idea of keying off name.

```js
"commands": {
  // ...
  { "name": "Split Horizontal", "action": "splitHorizontal", /* ... */ },
  // ...
},
"keybindings": {
  { "action": "executeCommand", "command": "Split Horizontal", ... }
}
```

What about localized names, if we support localization?

When you edit a command binding in the new UI, does it automatically update keybindings that hit it?

Not a huge fan, for the intricacies. Commands and Keybindings should be an abstraction over actions only.

Here's a thought, though: If we want to support _chained actions_ (one command does N things -- new tab, split horz, split vert, focus left)... should we just let the user define new _actions_, and use those in both commands AND keybinds?

Reply:
Hmm, this is good feedback. I'll pull that one. 

I like the idea of chained actions. Maybe we should have an action for it?
```json
{ 
    "action": "executeActions",
    "actions": [
        { "command": "newTab", "args": { ... } },
        { "command": "splitVertical" },
        { "command": "moveFocusLeft" }
    ]
}
```